iOS InContext Email Protection Signup

amddg44 opened the pull request:
<!--
Note: This checklist is a reminder of our shared engineering expectations. Feel free to change it, although assigning a GitHub reviewer and the items in bold are required.

⚠️ If you're an external contributor, please file an issue first before working on a PR, as we can't guarantee that we will accept your changes if they haven't been discussed ahead of time. Thanks!
-->

Task/Issue URL: https://app.asana.com/0/72649045549333/1202940260106121/f
Tech Design URL:
CC:

**Description**:
Adds Email InContext Signup protection to iOS

<!--
Tagging instructions
If this PR isn't ready to be merged for whatever reason it should be marked with the `DO NOT MERGE` label (particularly if it's a draft)
If it's pending Product Review/PFR, please add the `Pending Product Review` label.

If at any point it isn't actively being worked on/ready for review/otherwise moving forward (besides the above PR/PFR exception) strongly consider closing it (or not opening it in the first place). If you decide not to close it, make sure it's labelled to make it clear the PRs state and comment with more information.
-->

**Steps to test this PR**:
1. Starting with a clean install, visit `fill.dev/form/registration-email`
2. Confirm the grey dax icon is visible in the email field
3. Tap the text field (not the icon) and confirm the keyboard is displayed
4. Tap the icon and confirm the prompt to sign up for email protection is presented
5. Tap 'x' to dismiss the prompt and confirm the dax icon is still visible & grey
6. Tap the icon again, and this time select "Don't show again"
7. Confirm the dax icon is no longer visible and that refreshing the page does not bring it back
8. In the app go to Settings > Debug and tap the row "Reset Email Protection InContext Signup"
9. Return to `fill.dev/form/registration-email` and refresh the page. The grey dax icon should now be back
10. Tap the icon again, and this time select "Protect my email"
11. Choose "Get Started" to follow the flow to create a new duck address
12. Confirm that at the end of the sign up process, when dismissing the modal either from the 'Done' button or the 'Try it now', the prompt to chose a duck address is presented and that selecting an address populates the web form
13. Sign out and test the sign in flow (including tapping the link from email to launch the app)

<!--
Before submitting a PR, please ensure you have tested the combinations you expect the reviewer to test, then delete configurations you *know* do not need explicit testing.

Using a simulator where a physical device is unavailable is acceptable.
-->

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Orientation Testing**:

* [ ] Portrait
* [ ] Landscape

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [ ] iPhone 14 Pro
* [ ] iPad

**OS Testing**:

* [ ] iOS 14
* [ ] iOS 15
* [ ] iOS 16

**Theme Testing**:

* [ ] Light theme
* [ ] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

amddg44 committed:
Initial basic email promotion prompt modal flow

amddg44 committed:
Update to trigger autofill didRequestAliasAndRequiresUserPermission instead of calling openAutofillAfterClosingEmailProtectionTab

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into anya/ios-incontext-email-promotion

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj
#	DuckDuckGo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved

amddg44 committed:
Navigation styling & button state handling

amddg44 committed:
Flow update to trigger autofill replyHandler at end of incontext signup flow to trigger & populate duck address

amddg44 committed:
Improvement to email protection sign in flow when in the modal sign up overlay

amddg44 committed:
Reset of Email Protection InContext SignUp userDefaults added to app debug menu

amddg44 committed:
Email Protection InContext Signup strings added

amddg44 committed:
Pass in completion handler on init

amddg44 committed:
Blocking alert dialog for when attempting to exit the sign up flow early

amddg44 committed:
Replacing inContextSignup notification trigger with delegate method

amddg44 committed:
Removed un-needed delegate + some general tidy up

amddg44 committed:
New Email signup prompt view controller

amddg44 committed:
BSK version bump

Mcsadmeal reviewed (commented):
I

amddg44 committed:
Merge branch 'develop' into anya/ios-incontext-email-promotion

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj
#	DuckDuckGo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved
#	DuckDuckGo/Debug.storyboard
#	DuckDuckGo/RootDebugViewController.swift

amddg44 committed:
Reset Email InContext signup option added

amddg44 committed:
New Email address prompt

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into anya/ios-incontext-email-promotion

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj
#	DuckDuckGo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved
#	DuckDuckGo/TabViewController.swift

amddg44 committed:
Email prompt moved out of TabViewController.swift

amddg44 committed:
Update for SecureVaultStorage

amddg44 committed:
Design review feedback changes

amddg44 committed:
Set BSK, Autofill

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into anya/ios-incontext-email-promotion

amddg44 committed:
iPhone SE custom margin

amddg44 committed:
BSK bump

amddg44 committed:
BSK bump

amddg44 committed:
Layout constants

amddg44 committed:
BSK bump

amddg44 committed:
Tidy up

afterxleep commented on DuckDuckGo/AppDelegate.swift:
```diff
@@ -436,6 +436,14 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
     func application(_ app: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey: Any] = [:]) -> Bool {
         os_log("App launched with url %s", log: .lifecycleLog, type: .debug, url.absoluteString)
+
+        if url.absoluteString.starts(with: URL.emailProtection.absoluteString),
```
Nit.  WDYT about extracting this logic to a helper function or a url extenstion to keep this method clean?

afterxleep commented on DuckDuckGo/EmailSignupPromptViewController.swift:
```diff
@@ -0,0 +1,112 @@
+
+extension EmailSignupPromptViewController: UISheetPresentationControllerDelegate {
+    func presentationControllerDidDismiss(_ presentationController: UIPresentationController) {
+//            Pixel.fire(pixel: .autofillLoginsPasswordGenerationPromptDismissed)
```
Should this be removed?

afterxleep commented on DuckDuckGo/EmailSignupViewController.swift:
```diff
@@ -0,0 +1,466 @@
+                navigationItem.rightBarButtonItems = []
+                return
+            }
+            if url.absoluteString.hasSuffix("email/") || url.absoluteString.hasSuffix("email/start-incontext") {
```
These strings can be moved to your constants enum

afterxleep commented on DuckDuckGo/EmailSignupViewController.swift:
```diff
@@ -0,0 +1,466 @@
+
+    private lazy var backBarButtonItem: UIBarButtonItem = {
+        let button = UIButton(type: .system)
+        button.setImage(UIImage(systemName: "chevron.left"), for: .normal)
```
chevron.left can also be a constant

afterxleep commented on DuckDuckGo/EmailSignupViewController.swift:
```diff
@@ -0,0 +1,466 @@
+
+    private func setupNavigationBarTitle() {
+        let titleLabel: UILabel = UILabel()
+        titleLabel.text = "DuckDuckGo"
```
Should this be a localized string?  if not, probably a constant.

afterxleep commented on DuckDuckGo/EmailSignupViewController.swift:
```diff
@@ -0,0 +1,466 @@
+    @objc
+    private func onDuckDuckGoEmailSignIn(_ notification: Notification) {
+        if signupStage != .complete {
+            // TODO - pixel
```
Assuming this is what's coming up next with the pixels pr, right?

afterxleep commented on DuckDuckGo/AutofillLoginPromptView.swift:
```diff
@@ -93,6 +91,18 @@ struct AutofillLoginPromptView: View {
     }
 
+    private var horizontalPadding: CGFloat {
+        if AutofillViews.isIPhonePortrait(verticalSizeClass, horizontalSizeClass) {
```
I think the nested if can be replaced with a guard or even better a switch and a couple cases..

```
        guard AutofillViews.isIPhonePortrait.... else { 
            return Const.Size.closeButtonOffset 
         } 
        
        if isSmallFrame(frame) { 
            return ....
          } 
        
        return ....
```

afterxleep reviewed (approved):
Works as expected.  Left a few nitpicks

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into anya/ios-incontext-email-promotion

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj
#	DuckDuckGo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved
#	DuckDuckGo/MainViewController+Email.swift
#	DuckDuckGo/UserText.swift

amddg44 committed:
Updates from duck address changes merge

amddg44 committed:
Address PR feedback

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into anya/ios-incontext-email-promotion

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj
#	DuckDuckGo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved

amddg44 committed:
Bump BSK

amddg44 committed:
swiftlint

amddg44 committed:
Translations for InContext email protection (#1903)

Task/Issue URL: https://app.asana.com/0/0/1204945000199876/f
Tech Design URL:
CC:

Description:
Translations for in-context email protection

amddg44 committed:
iOS InContext Email Protection pixels (#1902)

Task/Issue URL: https://app.asana.com/0/0/1204874198702034/f
Tech Design URL:
CC:

Description:
Pixels for new inContext email protection prompt & sign in / up modal

amddg44 committed:
iOS InContext Email Protection promotion config updates (#1901)

Task/Issue URL: https://app.asana.com/0/72649045549333/1204936655327681/f
Tech Design URL:
CC:

Description:
Adds installedDays to privacy config to remotely control how many days since install that the iOS incontext email protection feature will be visible to users (currently set to less than 21 days). Also adds language restriction to limit feature to english speaking users

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into anya/ios-incontext-email-promotion

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj
#	DuckDuckGo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved
#	DuckDuckGo/Debug.storyboard
#	DuckDuckGo/TabViewController.swift

amddg44 committed:
BSK bump

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into anya/ios-incontext-email-promotion

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj
#	DuckDuckGo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved

amddg44 committed:
BSK bump

amddg44 committed:
BSK bump

amddg44 committed:
Set to BSK release 75.1.0
